### PR TITLE
Optimisations to intermediate variables

### PR DESCRIPTION
**Pull request by bensimner**
Also improves general optimisation performance, including running it to a fixed point, removing unobservable edges, allowing selection of individual optimisations to allow/remove.

### CODE REVIEW

**Comment by septract** [comment]
Builds and tests run for me.

File: Optimiser.fs [hunk 1]
```diff
@@ -680,6 +687,51 @@ module Graph =
                         else ctx
                 Set.fold processEdge ctx outEdges
 
+    /// Collapses edges {p}c{q}d{r} to {p}d{r} iff c is unobservable
```
Comment (septract) [inline Optimiser.fs:714]: More comment here about what 'unobservable' means?

Comment (bensimner) [inline Optimiser.fs:714]: Maybe something like

```
Collapses edges {p}c{q}d{r} to {p}d{r} iff c is uoberservable 
 i.e. c writes to local variables overwritten by d
 d does not read outputs of c, 
 and there are no assumes adding information
```

?

Comment (septract) [inline Optimiser.fs:714]: yes, great!

**Comment by septract** [comment]
This doesn't kick out as many inter variables as I'd expect. Eg. run: 

> `$ ./starling.sh -ssymproof -O all Examples/WIP/lclist-grasshopper.cvf`

I get the following clause: 

```
deleteVal_V005__deleteVal_V014_000:
    Command:
        (and
            (= (after lheap) (inter 5 heap))
            (= (inter 6 heap) (inter 5 heap))
            (= (inter 4 lheap) (inter 3 heap))
            (= (inter 4 heap) (inter 3 heap))
            (= (inter 2 lheap) (inter 1 heap))
            (= (inter 2 heap) (inter 1 heap))
            (= (inter 0 lheap) (before heap))
            (= (inter 1 lheap) (inter 0 lheap))
            (= (inter 1 heap) %{#2.next := #3.next }((inter 0 lheap), (before prev), (before curr)))
            (= (inter 3 lheap) (inter 2 lheap))
            (= (inter 3 heap) %{disposeNode(#2) }((inter 2 lheap), (before curr)))
            (= (inter 5 lheap) (inter 4 lheap))
            (= (inter 5 heap) %{#2.next := #3.next }((inter 4 lheap), (before prev), (before curr)))
            (= (after heap) %{disposeNode(#2) }((after lheap), (before curr)))
        )

    W/Prec:
        (and
            %{has1LockG(head,#1,#2) }((before prev), (before _))
            (= (goal 11 x) (before head))
            %{isListG(head,ub,#1) }((before v))
        )

    Goal:
        (= (goal 11 x) (before head))
```

I'd naively expect some of these to be kicked out, eg. `(= (inter 1 lheap) (inter 0 lheap))`.

**Comment by bensimner** [comment]
The problem with something like (= (inter 1 lheap) (inter 0 heap)) is that it's seeing it as (inter 1 lheap) := (inter 0 heap) and trying to apply the ordering. I'm betting we got the ordering the wrong way around previously (put a < when it should've been a >)

**Comment by bensimner** [comment]
Try that commit, I've flipped the ordering over inter -> inter assigns

It gives output of:

```
Command:
    (and
        (= (after lheap) (inter 5 heap))
        (= (inter 4 lheap) (inter 3 heap))
        (= (inter 2 lheap) (inter 1 heap))
        (= (inter 0 lheap) (before heap))
        (= (inter 1 heap) %{#2.next := #3.next }((inter 0 lheap), (before prev), (before curr)))
        (= (inter 3 heap) %{disposeNode(#2) }((inter 2 lheap), (before curr)))
        (= (inter 5 heap) %{#2.next := #3.next }((inter 4 lheap), (before prev), (before curr)))
        (= (after heap) %{disposeNode(#2) }((after lheap), (before curr)))
    )

W/Prec:
    (and
        %{has1LockG(head,#1,#2) }((before prev), (before _))
        (= (goal 11 x) (before head))
        %{isListG(head,ub,#1) }((before v))
    )

Goal:
    (= (goal 11 x) (before head))
```

It seems like there are still more variables containing intermediates that could be propogated perhaps, but they don't refer to the same name on both sides so aren't be optimised away here.

**Comment by septract** [comment]
Great, that looks like it works right on `lclist-grasshopper.cvf`.

**Comment by MattWindsor91** [comment]
@septract Is this ready to ship?  If so I can test and merge it.

**Comment by septract** [comment]
I'll take another look at the code today, but after that I think everything's good to go.

File: Optimiser.fs [hunk 2]
```diff
@@ -680,6 +687,54 @@ module Graph =
...
+
+                    let qViewexpr = (fun (viewexpr, _, _, _) -> viewexpr) <| ctx.Graph.Contents.[cEdge.Dest]
+                    let rViewexpr = (fun (viewexpr, _, _, _) -> viewexpr) <| ctx.Graph.Contents.[dEdge.Dest]
+
+                    match (pViewexpr, qViewexpr, rViewexpr) with
+                    | InnerView pView, InnerView qView, InnerView rView ->
+                        let cResults = Set.ofList <| commandResults c
+                        let dResults = Set.ofList <| commandResults d
+                        let dArgs    = commandArgs d
+                        if Set.isSubset cResults dResults
```
Comment (septract) [inline Optimiser.fs:741]: what's going on with assumes?

File: Optimiser.fs [hunk 3]
```diff
@@ -680,6 +687,54 @@ module Graph =
...
 
+    /// Collapses edges {p}c{q}d{r} to {p}d{r} iff c is uoberservable
+    /// i.e. c writes to local variables overwritten by d
+    /// d does not read outputs of c,
+    /// and there are no assumes adding information
+    let collapseUnoberservableEdges locals ctx =
+        expandNodeIn ctx <|
+            fun node nViewexpr outEdges inEdges nodeKind ->
+                let pViewexpr = nViewexpr
+                let disjoint a b = Set.empty = Set.filter (fun x -> Set.contains x b) a
```
Comment (septract) [inline Optimiser.fs]: better names here?

Comment (septract) [inline Optimiser.fs]: bracketing plz?

File: Optimiser.fs [hunk 4]
```diff
@@ -742,6 +797,7 @@ module Graph =
                      [ ("graph-collapse-nops", true, collapseNops)
                        ("graph-collapse-ites", true, collapseITEs)
                        ("graph-drop-local-edges", true, dropLocalEdges model.Locals)
+                       ("graph-collapse-unobservable-edges", true, collapseUnoberservableEdges model.Locals)
```
Comment (septract) [inline Optimiser.fs]: So if I change `true` -> `false` here, does that turn off the opt by default?

File: Utils.fs [hunk 5]
```diff
@@ -280,6 +280,12 @@ let foldFastTerm (f : 'State -> 'T -> 'State option)
 
     fft f s
 
+/// Repeatedly apply f until a fixed point is reached
```
Comment (septract) [inline Utils.fs:283]: Should perhaps pass a bound here and throw an error if it's exceeded? This seems like a place it'd be easy to accidentally introduce nontermination.

**Comment by septract** [comment]
Okay, this looks good modulo my comments and esp. the handling of assumes.

File: Optimiser.fs (hunk 4 above)
Comment (bensimner) [inline Optimiser.fs]: the behaviour is the same as @CaptainHayashi had originally (hopefully!)

As far as I understand, you can imagine these booleans as the "default" behaviour (enabled if true/disabled if false) and the command line arguments override this default.

File: Utils.fs (hunk 5 above)
Comment (bensimner) [inline Utils.fs:283]: Probably, what should the bound be?

Comment (septract) [inline Utils.fs:283]: dunno, 1000 or something? High enough that it's probably a loop if we ever see it. NB: there may be a less dumb way of doing this.

File: Optimiser.fs (hunk 4 above)
Comment (septract) [inline Optimiser.fs]: Okay, could you add a comment to this effect?

**Comment by bensimner** [comment]
Now passes all tests (if that means anything [it doesn't]) and regress.sh

**Comment by septract** [comment]
Looks okay, and passes regress.sh / NUnit. I think we're good to go unless @CaptainHayashi wants to take another look?

File: Command.fs [hunk 6]
```diff
@@ -68,6 +68,13 @@ module Queries =
     let commandResults cs =
         List.fold (fun a c -> a @ c.Results) [] cs
 
+    /// Retrieves the type annotated vars from the arguments to a
+    /// command as a list
+    let commandArgs cmd =
+        let f c = List.map SMExprVars c.Args
+        let vars = List.fold (@) [] <| List.map f cmd
```
Comment (MattWindsor91) [inline Command.fs:75]: Why not `List.concat`?  We also have `concatMap` in `Utils.fs`, which may be the right thing here.

File: Optimiser.fs [hunk 7]
```diff
@@ -481,6 +493,15 @@ module Graph =
                                               | _ -> false))
                      ps)
 
+    /// Decides whether a given Command contains any `assume` command
+    /// in any of the sequentially composed primitives inside it
+    let hasAssume : Command -> bool =
+        fun c ->
```
Comment (MattWindsor91) [inline Optimiser.fs:509]: Could do `hasAssume (c : Command) : bool =` here, or would this be stylistically bad?

File: Optimiser.fs [hunk 8]
```diff
@@ -481,6 +493,15 @@ module Graph =
...
+            c |>
+            List.forall (fun p ->
+                match p with
```
Comment (MattWindsor91) [inline Optimiser.fs]: Could replace this with `List.forall (function |...`

File: Optimiser.fs [hunk 9]
```diff
@@ -680,6 +701,57 @@ module Graph =
                         else ctx
                 Set.fold processEdge ctx outEdges
 
+    /// Collapses edges {p}c{q}d{r} to {p}d{r} iff c is uoberservable
+    /// i.e. c writes to local variables overwritten by d
+    /// d does not read outputs of c,
+    /// and there are no assumes adding information
+    let collapseUnoberservableEdges locals ctx =
```
Comment (MattWindsor91) [inline Optimiser.fs]: Typo?  Should be 'UnobservableEdges'

File: Symbolic.fs [hunk 10]
```diff
@@ -353,6 +353,33 @@ module Pretty =
     /// Pretty-prints a SMBoolExpr.
     let printSMBoolExpr = printBoolExpr (printSym printMarkedVar)
 
+/// Strip the marked part of the annotation
+/// and return just the internal 'var
+let unmarkMarkedVar =
+    function
+    | Before s            -> s
```
Comment (MattWindsor91) [inline Symbolic.fs:360]: Huh, I thought there was a function somewhere that did this already...

Comment (bensimner) [inline Symbolic.fs:360]: That one is a bijection, this is a surjection that just pulls the 'var out of it, if i recall correctly.

File: Optimiser.fs (hunk 9 above)
Comment (bensimner) [inline Optimiser.fs]: Yes

File: Optimiser.fs (hunk 7 above)
Comment (bensimner) [inline Optimiser.fs:509]: I've tried to keep a style

```
name : A -> B -> C =
    fun a b -> c
```

As I think it makes it clearer (specification then implementation), and fits the very Haskell-like codebase and use of types better ( i think ).

File: Optimiser.fs (hunk 8 above)
Comment (bensimner) [inline Optimiser.fs]: I could! I forget that F# has multiple ways of defining an anonymous function.

File: Command.fs (hunk 6 above)
Comment (bensimner) [inline Command.fs:75]: Good idea, I just didn't think, Obviously I'm still stuck in the world of maps and folds.

File: Optimiser.fs [hunk 11]
```diff
@@ -179,41 +179,53 @@ module Utils =
     /// </param>
     /// <returns>
     ///     A sequence of optimisers to run.
+    ///
+    ///     This builds the optimisers starting with all enabled
+    ///     removes everything that's disabled
+    ///     then adds everything that's enabled
     /// </returns>
```
Comment (MattWindsor91) [inline Optimiser.fs:184]: Nitpick: this should probably be in `<para>` tags.

File: Optimiser.fs [hunk 12]
```diff
@@ -179,41 +179,53 @@ module Utils =
...
+            if Set.contains "all" removes then
+                if config.verbose then
+                    eprintfn "note: all optimisations disabled"
+                Set.empty
+            elif Set.contains "all" adds then
+                    if config.verbose then
+                        eprintfn "note: all optimisations enabled"
+                    all_opts
+            else
+                /// backwards compatability (no args == all)
```
Comment (MattWindsor91) [inline Optimiser.fs]: This isn't a doc comment, so should be two slashes :P
Also, is this true?  Ideally if there is no `all` in either set then it should proceed to use the default lists of optimisations defined further down.

Comment (MattWindsor91) [inline Optimiser.fs]: And if there is an `all` in the `+` set then it should enable all optimisations without filtering (currently it's undefined behaviour to have `all` in both).

File: Optimiser.fs [hunk 13]
```diff
@@ -180,40 +182,60 @@ module Utils =
...
+            for (optName, _, _) in all_opts do
+                if optName.StartsWith(prefix) then
+                    if config.verbose && optimisationSet.Contains(optName) then
+                        eprintfn "note: forced %s off" optName
+
+                    ignore <| optimisationSet.Remove(prefix)
+
+        for (optName, enabledByDefault, _) in all_opts do
+            if enabledByDefault then
+                ignore <| optimisationSet.Add(optName)
```
Comment (MattWindsor91) [inline Optimiser.fs:207]: Could be

``` fsharp
List.iter (fun (optName, enabledByDefault, _) -> if enabledByDefault then optimisationSet.Add(optName))
```

I think, but it doesn't really matter.

File: Utils.fs [hunk 14]
```diff
@@ -281,10 +281,11 @@ let foldFastTerm (f : 'State -> 'T -> 'State option)
     fft f s
 
 /// Repeatedly apply f until a fixed point is reached
+let fix_bound = ref -1
```
Comment (MattWindsor91) [inline Utils.fs:284]: Global variable?  Really!?  :(

Comment (bensimner) [reply Utils.fs:284]: Yes.

File: Utils.fs [hunk 15]
```diff
@@ -297,3 +298,21 @@ module Testing =
...
+
+        [<Test>]
+        let ``test fix bounding`` () =
+            // check that (inc (inc (inc (0)))) = 3
+            check 3 3
+
+        [<Test>]
+        let ``test fix fix-point`` () =
+            // check that inc (inc (inc (0))) = 3
+            Assert.AreEqual (3, fix (cnst 3) 0)
```
Comment (MattWindsor91) [inline Utils.fs:318]: Woo, tests

**Comment by MattWindsor91** [comment]
:shipit:

**Pull request merged**